### ISSUE TITLE
external_storage dupe bug(1.12)

### ISSUE THREAD

**Issue by CatAndHorse** [raoulvdberge/refinedstorage#1407]
1.build an rs system with 2 external storage bus connect to 2 different chest
2.put 32 item into each chest
3.try to get 64 item by shift clicking in terminal or by export bus
4.you get 64 items and 32 remain  in rs sys

#### Version (make sure you are on the latest version before reporting):

- Minecraft: 1.12
- Forge:2420
- Refined Storage:1513

Does this issue occur on a server? [yes/no]
yes

**Comment by roboderpy** [comment]
Are you using Colossal chests? If so, are you connecting External Storage to Chest core?

**Comment by CatAndHorse** [comment]
Nope ,i tried vanilla chest , extrautil tiny chest and acatully addition wooden crate,all of them can reproduce this dupe

**Comment by raoulvdberge** [comment]
I can't reproduce this.
![x](https://raoulvdberge.com/PAp6Mo)

I have yet to test it on a server, though.

**Comment by CatAndHorse** [comment]
![bug](https://user-images.githubusercontent.com/4027330/29075237-5bcab346-7c84-11e7-8b35-2bbb426fb2dc.gif)


need 1 chest from mod and have higher prority

tested extrautils small chest and actually addons crate they both can reproduce this

**Comment by roboderpy** [comment]
@raoulvdberge confirmed.

This bug **can** be reproduced. @CatAndHorse just forgot to mention about **priority**.

**Comment by roboderpy** [comment]
But, as far as i can see it **can be reproduced only with one-slot inventories**. It just doesn't happen with regular chests only nor quark chests. Only Extra Utils mini chest can trigger that bug.
![2017-08-09_07 52 46](https://user-images.githubusercontent.com/6339511/29100777-10a74704-7cd8-11e7-9227-c59c8015fe7c.png)
![2017-08-09_07 52 47](https://user-images.githubusercontent.com/6339511/29100735-d3e5294e-7cd7-11e7-8e44-4994a1236324.png)
![2017-08-09_07 52 50](https://user-images.githubusercontent.com/6339511/29100736-d3e8e034-7cd7-11e7-9d11-f1506ef5aa97.png)
![2017-08-09_07 52 52](https://user-images.githubusercontent.com/6339511/29100737-d3e9c210-7cd7-11e7-8d90-fae708d7b92d.png)
![2017-08-09_07 53 01](https://user-images.githubusercontent.com/6339511/29100740-d3ee325a-7cd7-11e7-8ad4-d0d963c87869.png)
![2017-08-09_07 53 05](https://user-images.githubusercontent.com/6339511/29100738-d3ecfe8a-7cd7-11e7-85c5-ccc23c9de24b.png)
![2017-08-09_07 53 08](https://user-images.githubusercontent.com/6339511/29100739-d3ed7676-7cd7-11e7-9dad-ef39bb15a951.png)

**Comment by raoulvdberge** [comment]
In other words: this only happens with ExU mini chest? In that case, it might be an ExU dupe bug issue. Will still investigate.

**Comment by raoulvdberge** [comment]
This appears to be a ExU issue.

I tested this with 2 external storages both connected to its own Mini-Chest.

The problem is as follows:

- When shift clicking, RS asks 64 of that item
- Then, ExU gives back a stack of 64 items even though it only has 32

Proof:
![x](https://raoulvdberge.com/udNez2)

- `requested` is 64, because we want 64 items for shift click
- We got back a stack of 64 even though the Mini-Chest only has 32 of that item (look at `took`)

My testing setup:

![x](https://raoulvdberge.com/udNez2)

Both Mini-Chests have 32 items. But ExU just gives back 64 items for the first chest, even though it only has 32. So RS stops and doesn't extract from the second one.

In conclusion: this is a ExU issue. They appear to trust the "amount" parameter without checking if it doesn't exceed their own stack size.

**Comment by roboderpy** [comment]
And the problem is - @rwtema never reads github issues... ExtraUtils 2 issue tracker just drowned in issues

**Comment by raoulvdberge** [comment]
I just pinged him on IRC, he might fix it.

**Comment by CatAndHorse** [comment]
Actually additions' storage crates can reproduce this too @raoulvdberge 
okay i will open issue there too

**Comment by raoulvdberge** [comment]
I just looked into again some more and I was wrong: this is not an ExU or ActAdd issue. It's in fact an RS issue.

What's going on: I assumed `IItemHandler#extract` is guaranteed to make copies of the stack, and I could just safely modify the stack returned by it.

Apparantly it *does* copy the stack, but not when requesting more than the stack size.
```java
    @Nonnull
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (amount == 0) {
            return ItemStack.EMPTY;
        } else {
            this.validateSlotIndex(slot);
            ItemStack existing = (ItemStack)this.stacks.get(slot);
            if (existing.isEmpty()) {
                return ItemStack.EMPTY;
            } else {
                int toExtract = Math.min(amount, existing.getMaxStackSize());
                if (existing.getCount() <= toExtract) {
                    if (!simulate) {
                        this.stacks.set(slot, ItemStack.EMPTY);
                        this.onContentsChanged(slot);
                    }

                    return existing;
                } else {
                    if (!simulate) {
                        this.stacks.set(slot, ItemHandlerHelper.copyStackWithSize(existing, existing.getCount() - toExtract));
                        this.onContentsChanged(slot);
                    }

                    return ItemHandlerHelper.copyStackWithSize(existing, toExtract);
                }
            }
        }
    }
```
Look at the `return existing;`.
Thanks to that, RS just modifies the stack when extracting and false assumptions are
